src/metadata.rs

```diff
@@ -62,7 +63,10 @@ pub fn fetch_metadata(provider: &str) -> errors::Result<Box<dyn providers::Metad
         "ibmcloud" => box_result!(IBMGen2Provider::try_new()?),
         // IBM Cloud - Classic infrastructure.
         "ibmcloud-classic" => box_result!(IBMClassicProvider::try_new()?),
-        "openstack-metadata" => box_result!(OpenstackProvider::try_new()?),
+        "openstack" => match OpenstackConfigDrive::try_new() {
+            Ok(result) => box_result!(result),
+            Err(_) => box_result!(OpenstackProviderNetwork::try_new()?),
+        },
```

Comment:
I think this can be written as a simple `or_else` (but beware of matching types). Also, can you please:
 *  log a warning with the error from config-drive probing
 * move this small `try_autodetect()` logic to `openstack/mod.rs`

Reply:
Moved to `openstack/mod.rs` and added a `warn!("failed to utilize config-drive, using the metadata service API instead");` after config-drive fails and before using metadata service API.